### PR TITLE
Improve player connection and disconnection logic

### LINKED ISSUE #249: Possible Memory Leak
### Description of Issue
Bot fails at doing any task over Rcon due to potential Memory leak

Error:  (node:3891673) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 error listeners added to [Socket]. Use emitter.setMaxListeners() to increase limitError:  (Use `node --trace-warnings ...` to show where the warning was created)[LogParser][1] Lines parsed per minute: 807 lines per minute | Matching lines per minute: 186 matching lines per minute | Average matching latency: 1255.1290322580646ms

Output after above error:
[RCON][1] Unknown packet type 16777231 in: { size: 256, id: -1946157056, type: 16777231, body: '\x00\x00\x00\x00\x00\x00\x00----- Active Players -----\nID: 71 | SteamID: 76561199086558645 | Name: (   ÍœÍžÊ– ) | Team ID: 1 | Squad ID: 4 | Is Leader: False | Role: USA_Rifleman_03\nID: 75 | SteamID: 76561199259877501 | Name: 2945183**9 | Team ID: 1 | Squad ID: 4 | I' }

After above Unknown packet type output, no more Rcon log type 1 with player information is shown and below is looped.

[RCON][2] Writing packet with type "2" and body "ListPlayers".
[SquadServer][1] Updated A2S information.
[SquadServer][1] Updating player list...
[RCON][2] Writing packet with type "2" and body "ListPlayers".
[SquadServer][1] Updating player list...
[RCON][2] Writing packet with type "2" and body "ListPlayers".
[SquadServer][1] Updating player list...
[RCON][2] Writing packet with type "2" and body "ListPlayers".
[SquadServer][1] Updating player list...
[RCON][2] Writing packet with type "2" and body "ListPlayers".
[SquadServer][1] Updating player list...
[RCON][2] Writing packet with type "2" and body "ListPlayers".
[SquadServer][1] Updating player list...
[RCON][2] Writing packet with type "2" and body "ListPlayers".
[SquadServer][1] Updating player list...

### Squad Information
 * SquadJS Version: 3.4.0
 * Squad Version: 3.1.0

If potentially relevant, please provide regarding the state of the Squad server at the time of error, e.g. the current layer.

### System Information
 * Operating System: Ubuntu 20.04
 * NodeJS Version: 16.15.1
 * Yarn Version: 1.22.10

### PR DESCRIPTION
**Pull request by ect0s**
This is an initial port of changes from one of my branches

Fixes new EAC issues, provides a more reliable connection flow.
https://github.com/Team-Silver-Sphere/SquadJS/issues/257

Fixes Eventstore memory growth
https://github.com/Team-Silver-Sphere/SquadJS/issues/235

### CODE REVIEW

**Comment by ect0s** [comment]
unknown

Gist of the new player connection flow, I understand its hard to follow. 

Client Connected appears when a new network connection is made, and contains the steamid. 

We set this in eventstore.clients. These events are asyncronous from clients resolving. There is an error case where they get disconnected, pending connection destroyed. 

Next, we should see player client login, we use this to pull the steamid from the clients map, followed shortly by player controller connected, which sets up the player controller. 

Finally, player connected pulls the current steamid and player controller, plus player suffix, sets these up in eventStore.players, emits the finished event, and does some cleanup. 


Other events, particularly those around player damage, are stored in eventstore.sessions. 

This allows us to smartly clean these up on NEW GAME events, limiting overall memory growth within the logparser.

**Comment by ect0s** [comment]
Managed to finish some testing and this should be good.

File: squad-server/log-parser/player-connected.js [hunk 1]
```diff
@@ -6,9 +6,22 @@ export default {
       time: args[1],
       chainID: args[2],
       playerSuffix: args[3],
-      steamID: logParser.eventStore['steamid-connected']
+      steamID: logParser.eventStore['steamid-connected'], // player connected
```
Comment (Michael-Phoenix) [inline squad-server/log-parser/player-connected.js]: We removed steamid-connected in this PR
I think this should point towards client-login

File: squad-server/log-parser/player-connected.js [hunk 2]
```diff
@@ -6,9 +6,22 @@ export default {
...
+      controller: logParser.eventStore['player-controller'] // playercontroller connected
     };
 
+    delete logParser.eventStore['steamid-connected'];
```
Comment (Michael-Phoenix) [inline squad-server/log-parser/player-connected.js]: We removed steamid-connected in this PR
I think this should point towards client-login

**Review by Michael-Phoenix — CHANGES_REQUESTED** [review]
We removed steamid-connected in this PR
There are however still calls to the corresponding event store that is now empty

File: squad-server/log-parser/client-login.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+      time: args[1],
+      chainID: args[2],
+      connection: args[3]
+    };
+    /* This is Called when a player begins the Login process
+      We use this to get a SteamID into playerConnected.
+      2nd Step in player connected path
+      */
+
+    logParser.eventStore['steamid-connected'] = logParser.eventStore.clients[args[3]];
```
Comment (Michael-Phoenix) [inline squad-server/log-parser/client-login.js]: It's probably good to change the eventstore key to client-login for consistency.
steamid-connected is just legacy code at this point

**Review by Michael-Phoenix — CHANGES_REQUESTED** [review]

**Comment by Michael-Phoenix** [comment]
Thank you for this change! Love the pic to make the flow more understandable! 
Fantastic job!

File: squad-server/log-parser/player-connected.js (hunk 1 above)
Comment (werewolfboy13) [reply squad-server/log-parser/player-connected.js]: Planning on keeping this as is for now. Trying to play catch up with the PRs. If you want to add these changes in you are welcome to add your on PR for it.

File: squad-server/log-parser/player-connected.js (hunk 2 above)
Comment (werewolfboy13) [reply squad-server/log-parser/player-connected.js]: Planning on keeping this as is for now. Trying to play catch up with the PRs. If you want to add these changes in you are welcome to add your on PR for it.

File: squad-server/log-parser/client-login.js (hunk 3 above)
Comment (werewolfboy13) [reply squad-server/log-parser/client-login.js]: Planning on keeping this as is for now. Trying to play catch up with the PRs. If you want to add these changes in you are welcome to add your on PR for it.

**Review by werewolfboy13 — DISMISSED** [review]
Tested long term on a live server. No issues reported.

File: core/log-parser/index.js [hunk 4]
```diff
@@ -14,7 +14,12 @@ export default class LogParser extends EventEmitter {
 
     options.filename = filename;
 
-    this.eventStore = {};
+    this.eventStore = {
+      disconnected: {}, // holding area, cleared on map change.
+      players: {}, // persistent data, steamid, controller, suffix.
+      session: {}, // old eventstore, nonpersistent data
```
Comment (Thomas-Smyth) [inline core/log-parser/index.js:20]: At some point, we should split these out to separate objects as well.

Not critical for this PR though.

File: core/log-parser/index.js [hunk 5]
```diff
@@ -61,6 +66,19 @@ export default class LogParser extends EventEmitter {
...
+  clearEventStore() {
+    Logger.verbose('LogParser', 2, 'Cleaning Eventstore');
+    for (const player of Object.values(this.eventStore.players)) {
+      if (this.eventStore.disconnected[player.steamID] === true) {
+        Logger.verbose('LogParser', 2, `Removing ${player.steamID} from eventStore`);
+        delete this.eventStore.players[player.steamID];
+        delete this.eventStore.disconnected[player.steamID];
+      }
+    }
+    this.eventStore.session = {};
```
Comment (Thomas-Smyth) [inline core/log-parser/index.js:79]: I don't think we currently clean up the event store. Have you tested what side effects this has, e.g. do these events always get triggered before the new game such as someone bleeding out after the end of the game? I'm not fussed about counting the bleed out event, but need to check that kind of thing doesn't cause any errors.

File: squad-server/log-parser/client-login.js (hunk 3 above)
Comment (Thomas-Smyth) [reply squad-server/log-parser/client-login.js]: I agree with @Michael-Phoenix on this, it should be a simple change to do and would ensure the code is clear in the long run - I doubt anyone will remember to change it in the future.

**Review by Thomas-Smyth — DISMISSED** [review]
Looks good to me, and I gather this has been tested long-term in production.

A few minor questions and some quick fixes to apply, but otherwise looks good to merge.

Ah, could you also rename the PR to something suitable for the changelog please?

File: core/log-parser/index.js (hunk 4 above)
Comment (ect0s) [reply core/log-parser/index.js:20]: This PR is heavily cutdown version of a log parser change I tested with BB | wasp, where on startup squadjs scans the entire log file and builds up these caches/syncs state. There are issues around that PR and how it was built, but 
I'm testing a version where the eventstore actually ends up as a cache in the server object. 

As is, currently no plugins really go back and touch this information, and our information in server.players gets refreshed every cycle; but we are not "attaching" the log specific entries here over there. 

In my branch, I'm using the logparser cache to do lookups by player controller directly in damage events and so on.

We also end up with a way of tracking sessions directly on each player, but thats also possible elsewhere.

File: core/log-parser/index.js (hunk 5 above)
Comment (ect0s) [reply core/log-parser/index.js:79]: https://github.com/Team-Silver-Sphere/SquadJS/blob/17a7c269a7e65013d314bed2f5b56d2a0179409b/squad-server/log-parser/new-game.js#L21

We call the cleanup on newgame, which is triggered on the map change. At this point, all players have been pushed to scoreboard and no more events from damage/kills/wounds are generated. 

Since this cache is self contained within the log parser, it should be self consistent, however, large FTP log parser latencies may cause issues within the server class.

Comment (ect0s) [reply core/log-parser/index.js:79]: The reason for clearing .session every game, is in current/old squadjs, particularly around wounds/damage events, especially when they contain "nullptr" in the regex, is we may be assembling events from data spanning two rounds unintentionally. 

There are alot of edge cases in how damage/wound/died is generated and how we cycle forward attaching information, and I know for a fact this has lead to incorrectly attributed data coming out of the log parser in the current revision. 

While I did beat my head for a while working on a more structured fix, in the end I'm not sure how to deal with ambiguous cases and so left this as is for now.

Comment (Thomas-Smyth) [reply core/log-parser/index.js:79]: > At this point, all players have been pushed to scoreboard and no more events from damage/kills/wounds are generated.

Is this something that's changed? I thought previous tickets were still updated when the scoreboard was loaded as people were bleeding out during that time, etc.

Comment (ect0s) [reply core/log-parser/index.js:79]: As for why I keep disconnected and players separate and only run the cleanup on new game:

My concern was probably what your pointing to: If we have a bleedout after a player leaves, we won't have the disconnected player within the cache.

The solution was to *keep* the players within the cache until the round ends. 

So, in practice,A new round starts:

Player A downs Player B
Player A Disconnects. 

Player B bleeds out (we attach player A).

New game fires, and we cleanup player A. 

I admit this is a very "manual" solution, but it fixes the "life time" issues via manual "garbage" collection. If someone has a more elegant solution feel free to let me know.

File: core/log-parser/index.js (hunk 4 above)
Comment (Thomas-Smyth) [reply core/log-parser/index.js:20]: It's not critical for this PR, so this can be resolved. However, my chain of thoughts is currently everything in that object is keyed by just a player's name, so if we start adding more events that overlap then data could be overwritten unintentionally. If all the data in this object just related to damage/wound/death/revive that's fine, we just need to apply a more useful name for the object.

File: core/log-parser/index.js (hunk 5 above)
Comment (ect0s) [reply core/log-parser/index.js:79]: > Is this something that's changed? I thought previous tickets were still updated when the scoreboard was loaded as people were bleeding out during that time, etc.

So, when we hit the scoreboard, all players "unposses" their objects, or should from testing. 

```
[2022.12.29-09.08.22:442][813]LogGameMode: Display: Match State Changed from InProgress to WaitingPostMatch
[2022.12.29-09.08.22:442][813]LogSquadTrace: [DedicatedServer]OnUnPossess(): PC=Doug1952
[2022.12.29-09.08.22:442][813]LogSquadTrace: [DedicatedServer]OnUnPossess(): PC=Doug1952 current health value 5.000000
[2022.12.29-09.08.22:442][813]LogSquadTrace: [DedicatedServer]ChangeState(): PC=Doug1952 OldState=Playing NewState=Inactive
[2022.12.29-09.08.22:442][813]LogSquadTrace: [DedicatedServer]BeginInactiveState(): PC=Doug1952
[2022.12.29-09.08.22:443][813]LogSquadTrace: [DedicatedServer]OnUnPossess(): PC=Tmr539
[2022.12.29-09.08.22:443][813]LogSquadTrace: [DedicatedServer]OnUnPossess(): PC=Tmr539 current health value -9.223495
[2022.12.29-09.08.22:443][813]LogSquadTrace: [DedicatedServer]ChangeState(): PC=Tmr539 OldState=Playing NewState=Inactive
[2022.12.29-09.08.22:443][813]LogSquadTrace: [DedicatedServer]BeginInactiveState(): PC=Tmr539
[2022.12.29-09.08.22:444][813]LogSquadTrace: [DedicatedServer]OnUnPossess(): PC=Slipped DabSloth
[2022.12.29-09.08.22:444][813]LogSquadTrace: [DedicatedServer]OnUnPossess(): PC=Slipped DabSloth current health value 35.220371
[2022.12.29-09.08.22:444][813]LogSquadTrace: [DedicatedServer]ChangeState(): PC=Slipped DabSloth OldState=Playing NewState=Inactive
[2022.12.29-09.08.22:444][813]LogSquadTrace: [DedicatedServer]BeginInactiveState(): PC=Slipped DabSloth

....


[2022.12.29-09.08.22:494][813]LogSquadGameEvents: Display: Team 1, US Brigade Combat Team ( United States Army ) has won the match with 375 Tickets on layer Narva AAS v3 (level Narva)!
[2022.12.29-09.08.22:494][813]LogSquadGameEvents: Display: Team 2, Russian battalion tactical group ( Russian Ground Forces ) has lost the match with 0 Tickets on layer Narva AAS v3 (level Narva)!
[2022.12.29-09.08.22:494][813]LogGameState: Match State Changed from InProgress to WaitingPostMatch

```

In theory, the cleanup *could* be called at 

```
LogGameState: Match State Changed from InProgress to WaitingPostMatch
```

Which is used by the logtickets PR

Comment (Thomas-Smyth) [reply core/log-parser/index.js:79]: I presume once unpossessed they are cleaned up, or at least their deaths, etc. aren't logged?

Comment (ect0s) [reply core/log-parser/index.js:79]: They don't count as "deaths" from the perspective of player-died.

IE, we won't see the regex for player died between the above log and the next newgame.

Comment (ect0s) [reply core/log-parser/index.js:79]: In testing, the above is 100% the case, 

```
PLAYER_UNPOSSESS 2022-12-29T09:08:22.442Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.442Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.443Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.443Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.444Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.444Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.445Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.445Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.446Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.447Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.447Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.448Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.448Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.449Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.449Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.450Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.450Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.451Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.453Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.453Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.454Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.454Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.455Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.455Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.456Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.456Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.457Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.457Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.458Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.458Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.458Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.458Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.459Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.459Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.460Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.460Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.461Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.461Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.462Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.462Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.463Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.463Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.463Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.463Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.464Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.464Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.465Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.465Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.466Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.466Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.467Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.467Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.468Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.468Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.469Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.469Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.470Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.470Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.471Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.471Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.471Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.471Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.472Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.472Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.473Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.473Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.474Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.474Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.475Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.475Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.476Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.476Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.477Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.477Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.477Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.477Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.478Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.478Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.479Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.479Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.480Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.480Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.481Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.482Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.485Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.485Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.485Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.485Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.486Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.486Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.487Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.487Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.488Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.488Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.489Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.489Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.490Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.490Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.491Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.491Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.492Z
PLAYER_UNPOSSESS 2022-12-29T09:08:22.492Z
PLAYER_UNPOSSESS 2022-12-29T09:08:26.921Z
PLAYER_UNPOSSESS 2022-12-29T09:08:34.987Z
PLAYER_UNPOSSESS 2022-12-29T09:08:37.683Z
PLAYER_UNPOSSESS 2022-12-29T09:08:39.354Z
PLAYER_UNPOSSESS 2022-12-29T09:08:39.421Z
PLAYER_UNPOSSESS 2022-12-29T09:08:40.691Z
PLAYER_UNPOSSESS 2022-12-29T09:08:41.448Z
PLAYER_UNPOSSESS 2022-12-29T09:08:43.164Z
PLAYER_UNPOSSESS 2022-12-29T09:08:44.166Z
PLAYER_UNPOSSESS 2022-12-29T09:08:45.214Z
PLAYER_UNPOSSESS 2022-12-29T09:08:45.682Z
PLAYER_UNPOSSESS 2022-12-29T09:08:47.152Z
PLAYER_UNPOSSESS 2022-12-29T09:08:49.092Z
NEW_GAME Fallujah_RAAS_v7 2022-12-29T09:08:54.391Z
```

Excuse the formatting, this was done offline

Comment (ect0s) [reply core/log-parser/index.js:79]: [out.zip](https://github.com/Team-Silver-Sphere/SquadJS/files/10329688/out.zip)

Comment (ect0s) [reply core/log-parser/index.js:79]: On tickets, it may be the case that players that have negative health values (but not -300) effect tickets when they are cleaned up as above, but this is an issue I'm not sure we can work around. 

This *could* by why sometimes we see tickets adjusting at the scoreboard, but I'm not entirely sure. I think players leaving also may have an effect here, but these are squad bugs, not squadjs bugs. 

As for renaming the PR, I can do that. 

Not sure whats appropriate since this rolls a few things together;

**Comment by ect0s** [comment]
In the process of renaming steamid-connected I had to roll forward to master branch to deal with merge conflicts

But this *should* be good to go
